taskfiles/utils.yml

```diff
@@ -141,3 +146,194 @@ tasks:
         \( -iname "*.cpp" -o -iname "*.h" -o -iname "*.hpp" \) \
         -print0 | \
           xargs -0 clang-tidy {{.FLAGS}}
+
+  # ===
+  # CMAKE UTILS
+  # ===
+
+  # Runs cmake configure and build steps for the given source and build directories.
+  #
+  # @param {string} BUILD_DIR Cmake build directory to create.
+  # @param {string} SOURCE_DIR Project source directory containing the CMakeLists.txt file.
+  # @param {string={{.BUILD_DIR}}.md5} [CHECKSUM_FILE] Path to store the checksum of built files.
+  # @param {string=""} [CMAKE_ARGS] Any additional arguments to pass to cmake configure.
+  cmake-build:
+    label: "{{.TASK}}-{{.SOURCE_DIR}}-{{.BUILD_DIR}}"
+    internal: true
+    vars:
+      CHECKSUM_FILE: >-
+        {{default (printf "%s.md5" .BUILD_DIR) .CHECKSUM_FILE}}
+      CMAKE_ARGS: >-
+        {{default "" .CMAKE_ARGS}}
+    requires:
+      vars: ["BUILD_DIR", "SOURCE_DIR"]
+    sources:
+      - "{{.SOURCE_DIR}}/**/*"
+      - exclude: "{{.SOURCE_DIR}}/.cache/**/*"
+      - exclude: "{{.SOURCE_DIR}}/compile_commands.json"
+    generates: ["{{.CHECKSUM_FILE}}"]
+    deps:
+      - task: "validate-checksum"
+        vars:
+          CHECKSUM_FILE: "{{.CHECKSUM_FILE}}"
+          DATA_PATHS: ["{{.BUILD_DIR}}"]
+          EXCLUDE_PATHS: ["install_manifest.txt"]
+    cmds:
+      - >-
+        cmake
+        -S "{{.SOURCE_DIR}}"
+        -B "{{.BUILD_DIR}}"
+        {{.CMAKE_ARGS}}
+      - >-
+        cmake
+        --build "{{.BUILD_DIR}}"
+        --parallel
+      # This command must be last
+      - task: "compute-checksum"
+        vars:
+          DATA_PATHS: ["{{.BUILD_DIR}}"]
+          OUTPUT_FILE: "{{.CHECKSUM_FILE}}"
+          EXCLUDE_PATHS: ["install_manifest.txt"]
+
+  # Runs cmake install step for the given build directory.
+  #
+  # @param {string} BUILD_DIR Cmake build directory.
+  # @param {string} INSTALL_PREFIX Path prefix for installing the project.
+  # @param {string={{.INSTALL_PREFIX}}.md5} [CHECKSUM_FILE] Path to store the checksum of installed
+  # files.
+  # @param {string[]=[{{.INSTALL_PREFIX}}]} [DATA_PATHS] Paths to compute the the checksum for.
```

Comment:
```suggestion
  # @param {string[]=[{{.INSTALL_PREFIX}}]} [DATA_PATHS] Paths to compute the checksum for.
```

Reply:
I might be missing something, but this is what I've been going off of:

Without allowing the caller to specify the install prefix, we would always install to the system path by default. Installing to the system path could lead to conflicts or permission issues. The user could manually sets the `CMAKE_INSTALL_PREFIX` environment variable before running `task` but this would blindly apply to all `cmake-install` tasks.

For example: in ffi-go when we build the dependencies, I install them into the build directory (rather than the system) to ensure there are no conflicts and easy clean up.

ref: https://cmake.org/cmake/help/latest/variable/CMAKE_INSTALL_PREFIX.html

Reply:
Oh I'm fine with `INSTALL_PREFIX`---I mean what do we need `DATA_PATHS` for?

Reply:
Sorry too asleep when I responded lol.

We only want to compute the checksum for files we install, but a user might want to install to a file with other files. 
For example, say install prefix is `/usr/local` and we're installing library `foo`. We could end up with:
```
/usr/local/lib/bar
/usr/local/lib/foo
```

In a perfect world we could know what was installed from CMake (e.g. `lib/foo`), but that sounds hard :(.

Alternatively, I guess we could just always install and don't bother computing a checksum or using generates?